### PR TITLE
findFreePorts fix

### PR DESCRIPTION
**Pull request by arndey**

### CODE REVIEW

File: modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/Utils.kt [hunk 1]
```diff
@@ -4,32 +4,51 @@ import com.github.dockerjava.api.DockerClient
...
-        return (0 until amount).map {
-            var socket = ServerSocket(0)
-            while (socket.localPort in busyPorts) {
+        var socket: ServerSocket? = null
+        var retriesLeft = retries
+
+        while (retriesLeft > 0) {
+            try {
                 socket = ServerSocket(0)
+                while (socket?.localPort in busyPorts || socket?.localPort in taken) {
```
Comment (Mingela) [inline modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/Utils.kt:35]: some timeout maybe?

Comment (Mingela) [reply modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/Utils.kt:35]: what's the point of retries in that case

File: modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/Utils.kt [hunk 2]
```diff
@@ -4,32 +4,51 @@ import com.github.dockerjava.api.DockerClient
...
+                    socket = ServerSocket(0)
+                }
+                return PortToSocket(socket!!.localPort, socket).also { taken.add(it.port) }
```
Comment (Mingela) [inline modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/Utils.kt:38]: we need to guarantee a kind of atomicity of this, how about using AtomicReference<ServerSocket> or smth

File: modules/test-tools/src/main/kotlin/jp/co/soramitsu/iroha2/testengine/IrohaConfig.kt [hunk 3]
```diff
@@ -38,6 +39,7 @@ class IrohaConfig(
         PeerId(SocketAddr.Host(SocketAddrHost(alias, DEFAULT_P2P_PORT)), keyPair.public.toIrohaPublicKey()),
     ),
     var ports: List<Int> = listOf(DEFAULT_P2P_PORT, DEFAULT_API_PORT, DEFAULT_TELEMETRY_PORT),
+    var sockets: List<ServerSocket> = emptyList(),
```
Comment (Mingela) [inline modules/test-tools/src/main/kotlin/jp/co/soramitsu/iroha2/testengine/IrohaConfig.kt:42]: what's the point, not used?

File: modules/test-tools/src/main/kotlin/jp/co/soramitsu/iroha2/testengine/IrohaRunnerExtension.kt [hunk 4]
```diff
@@ -131,6 +136,8 @@ class IrohaRunnerExtension : InvocationInterceptor, BeforeEachCallback {
             ).also { utilizedResources.add(it) }
         }
 
+        containers.map { utilizedResources.addAll(it.getSockets()) }
```
Comment (Mingela) [inline modules/test-tools/src/main/kotlin/jp/co/soramitsu/iroha2/testengine/IrohaRunnerExtension.kt:139]: what does this imply?

**Comment by Mingela** [comment]
Please set a clear PR name

File: modules/test-tools/src/main/kotlin/jp/co/soramitsu/iroha2/testengine/IrohaRunnerExtension.kt (hunk 4 above)
Comment (arndey) [reply modules/test-tools/src/main/kotlin/jp/co/soramitsu/iroha2/testengine/IrohaRunnerExtension.kt:139]: to close a socket after test

File: modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/Utils.kt [hunk 5]
```diff
@@ -4,32 +4,51 @@ import com.github.dockerjava.api.DockerClient
...
+            } catch (e: IOException) {
+                // Port not available
+            } finally {
+                socket?.close() // To let docker take it
```
Comment (Mingela) [inline modules/client/src/main/kotlin/jp/co/soramitsu/iroha2/Utils.kt:42]: let's maybe do that somehow when docker tries to take it, not here?

**Pull request closed without merging**